desc/objectives/_coils.py

```diff
@@ -96,102 +96,74 @@ def build(self, use_jit=True, verbose=1):  # noqa:C901
+        errorif(
+            np.any([g.num_rho > 1 or g.num_theta > 1 for g in grid]),
+            ValueError,
+            "Only use toroidal resolution for coil grids.",
         )
-        self._num_coils = len(coil_leaves)
-
-        # check type
-        if isinstance(self._grid, numbers.Integral):
-            self._grid = LinearGrid(N=self._grid, endpoint=False)
-        # all of these cases return a container MixedCoilSet that contains
-        # LinearGrids. i.e. MixedCoilSet.coils = list of LinearGrid
-        if self._grid is None:
-            # map default grid to structure of inputted coils
-            self._grid = tree_map(
-                lambda x: LinearGrid(
-                    N=2 * x.N + 5, NFP=getattr(x, "NFP", 1), endpoint=False
-                ),
-                self.things[0],
-                is_leaf=lambda x: not hasattr(x, "__len__"),
-            )
-        elif isinstance(self._grid, _Grid):
-            # map inputted single LinearGrid to structure of inputted coils
-            self._grid = [self._grid] * self._num_coils
-            self._grid = tree_unflatten(coil_structure, self._grid)
-        else:
-            # this case covers an inputted list of grids that matches the size
-            # of the inputted coils. Can be a 1D list or nested list.
-            flattened_grid = tree_leaves(
-                self._grid, is_leaf=lambda x: isinstance(x, _Grid)
-            )
-            self._grid = tree_unflatten(coil_structure, flattened_grid)
+
+        self._dim_f = np.sum([g.num_nodes for g in grid])
+        quad_weights = np.concatenate([g.spacing[:, 2] for g in grid])
 
         timer = Timer()
         if verbose > 0:
             print("Precomputing transforms")
         timer.start("Precomputing transforms")
 
+        # map grid/transform to the same structure as coil
+        grid = tree_unflatten(structure, grid)
         transforms = tree_map(
-            lambda x, y: get_transforms(self._data_keys, obj=x, grid=y),
-            self.things[0],
-            self._grid,
-            is_leaf=lambda x: not hasattr(x, "__len__"),
-        )
-
-        grids = tree_leaves(self._grid, is_leaf=lambda x: hasattr(x, "num_nodes"))
-        self._dim_f = np.sum([grid.num_nodes for grid in grids])
-        self._quad_weights = np.concatenate([grid.spacing[:, 2] for grid in grids])
-
-        # get only needed grids (1 per CoilSet) and flatten that list
-        self._grid = tree_leaves(
-            to_list(self._grid), is_leaf=lambda x: isinstance(x, _Grid)
-        )
-        transforms = tree_leaves(
-            to_list(transforms), is_leaf=lambda x: isinstance(x, dict)
-        )
-
-        errorif(
-            np.any([grid.num_rho > 1 or grid.num_theta > 1 for grid in self._grid]),
-            ValueError,
-            "Only use toroidal resolution for coil grids.",
+            lambda c, g: get_transforms(self._data_keys, obj=c, grid=g),
```

Comment:
minor point: This will get more transforms than needed, since the ones for `CoilSet` are later pruned to just the unique ones. Would be nice if we can avoid that redundant calculation but not sure how feasible it is.

Reply:
I think I've resolved this in the latest commit